### ISSUE TITLE
Detecting Two Objects with the Same Center

### ISSUE THREAD

**Issue by nadavkirsch-6over6** [ultralytics/yolov5#8068]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello,

I have encountered a problem while working with yolov5 on my custom data. The problem is - If there are two objects that have the same center point, the model fails to detect the inner object.
The model detects both objects with low confidence and in the exact same shape. 
I found that increasing the imgsz might help but still, the results are pretty bad.
I wanted to ask if there is a known solution to this problem or any ideas you might have since I didn't find any solutions.
I am attaching two images, one with the original bounding boxes, and the other with the predicted bounding boxes. In these images I am trying to detect the eye and the iris, and their bounding boxes share almost the exact same center point. 

Thank you!

![nadav](https://user-images.githubusercontent.com/97099126/171381503-c5d79976-937b-4f5b-9829-31348dd9b967.jpg)
![nadav](https://user-images.githubusercontent.com/97099126/171381568-f54ba23b-6d55-4d86-9132-e53975cbdab8.jpeg)

**Comment by glenn-jocher** [comment]
@nadavkirsch-6over6 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 


Identical centerpoints may result in ambiguity about which box to predict. Some architectures like FCOS attempt to solve this by configuring the prediction space differently, but before considering any exotic solutions I would review the recommendations below to verify your dataset is sufficiently aligned with best practices.
<img width="355" alt="Screen Shot 2022-06-01 at 2 15 41 PM" src="https://user-images.githubusercontent.com/26833433/171402272-6987d2b8-247c-493b-8180-af06d0f543b6.png">


Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by gillonba** [comment]
How important is it that... you only look once?  Would it be a problem to split the job into two models, run the detection with each, and then combine the results?  

Alternatively, since you appear to be detecting an eye in addition to a specific part of the eye, perhaps you could detect the eye, cut out the portion of the image in the detection, and then run your second stage detector on just the image of the eye?  It would be for more expensive computationally, but much simpler to see what is really happening.  Additionally it would be easy to collect a lot of source images of just eyes to build your iris dataset.  I am doing something similar to do different classifications on detected objects and am happy with the initial results

**Comment by glenn-jocher** [comment]
@gillonba great questions! 🤔 

The "only look once" nature of YOLO refers to its approach of predicting all bounding boxes and class probabilities from a single forward pass of the network, combining high levels of precision with real-time speeds. While running two models and combining results can work, it would indeed be more computationally expensive and time-consuming compared to a single, optimized model.

Your alternative approach of first detecting the entire eye and then running a second stage detector on just the eye image is a promising strategy! This two-stage detection method could potentially simplify the detection process and even improve the overall accuracy, while also allowing for easier dataset collection as you mentioned. 

Your results sound impressive! It's great to hear that you are happy with the initial outcomes. Keep up the great work and don't hesitate to reach out if you need further assistance!
